### ISSUE TITLE
[Gap Analysis] Implement RMSE and Sparse Categorical Cross-Entropy Loss

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#331]
### User Story

> As a machine learning engineer, I want access to a complete set of standard loss functions, including specific variants for common use cases like integer-encoded categorical labels, so that I can precisely configure my models for various tasks.

---

### **Problem: Minor Gaps in Loss Function Coverage**

The `src/LossFunctions` module is highly comprehensive. However, two standard loss functions, commonly found in other frameworks, are currently missing.

---

### **Phase 1: Implement Root Mean Squared Error (RMSE) Loss**

**Goal:** Add RMSE as a direct loss function, often used as a metric.

#### **AC 1.1: Create `RootMeanSquaredErrorLoss.cs` (5 points)**
**Requirement:** Implement the RMSE loss function.

-   [ ] **File:** `src/LossFunctions/RootMeanSquaredErrorLoss.cs`
-   [ ] **Class:** `public class RootMeanSquaredErrorLoss<T> : LossFunctionBase<T>`
-   [ ] **Methods:** `public override T Forward(Vector<T> predictions, Vector<T> targets)`, `public override Vector<T> Backward(Vector<T> predictions, Vector<T> targets)`.
-   [ ] **Logic:** Calculate the square root of the mean of squared differences between predictions and targets.

#### **AC 1.2: Unit Tests for `RootMeanSquaredErrorLoss` (3 points)**
**Requirement:** Verify the correctness of RMSE loss.

-   [ ] **File:** `tests/UnitTests/LossFunctions/RootMeanSquaredErrorLossTests.cs`
-   [ ] **Test Cases:** Test `Forward` and `Backward` passes with various predictions and targets.

---

### **Phase 2: Implement Sparse Categorical Cross-Entropy Loss**

**Goal:** Add a cross-entropy variant for integer-encoded categorical labels.

#### **AC 2.1: Create `SparseCategoricalCrossEntropyLoss.cs` (8 points)**
**Requirement:** Implement Sparse Categorical Cross-Entropy Loss.

-   [ ] **File:** `src/LossFunctions/SparseCategoricalCrossEntropyLoss.cs`
-   [ ] **Class:** `public class SparseCategoricalCrossEntropyLoss<T> : LossFunctionBase<T>`
-   [ ] **Methods:** `public override T Forward(Matrix<T> predictions, Vector<int> targets)`, `public override Matrix<T> Backward(Matrix<T> predictions, Vector<int> targets)`.
-   [ ] **Logic:** Similar to Categorical Cross-Entropy, but `targets` are integer class indices (not one-hot encoded).

#### **AC 2.2: Unit Tests for `SparseCategoricalCrossEntropyLoss` (5 points)**
**Requirement:** Verify the correctness of Sparse Categorical Cross-Entropy Loss.

-   [ ] **File:** `tests/UnitTests/LossFunctions/SparseCategoricalCrossEntropyLossTests.cs`
-   [ ] **Test Cases:** Test `Forward` and `Backward` passes with multi-class predictions and integer-encoded targets.

---

### **Definition of Done**

-   [ ] All checklist items are complete.
-   [ ] `RootMeanSquaredErrorLoss` and `SparseCategoricalCrossEntropyLoss` are implemented and unit-tested.
-   [ ] All new tests pass.
---

## ⚠️ CRITICAL ARCHITECTURAL REQUIREMENTS

**Before implementing this user story, you MUST review:**
- **📋 Full Requirements:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
- **📐 Project Rules:** [`.github/PROJECT_RULES.md`](../.github/PROJECT_RULES.md)

#### 1. INumericOperations<T> Usage (CRITICAL)
- [ ] Include `protected static readonly INumericOperations<T> NumOps = MathHelper.GetNumericOperations<T>();` in base class
- [ ] NEVER hardcode `double`, `float`, or specific numeric types - use generic `T`
- [ ] NEVER use `default(T)` - use `NumOps.Zero` instead
- [ ] Use `NumOps.Zero`, `NumOps.One`, `NumOps.FromDouble()` for values
- [ ] Use `NumOps.Add()`, `NumOps.Multiply()`, etc. for arithmetic
- [ ] Use `NumOps.LessThan()`, `NumOps.GreaterThan()`, etc. for comparisons

#### 2. Inheritance Pattern (REQUIRED)
- [ ] Create `I{FeatureName}.cs` in `src/Interfaces/` (root level, NOT subfolders)
- [ ] Create `{FeatureName}Base.cs` in `src/{FeatureArea}/` inheriting from interface
- [ ] Create concrete classes inheriting from Base class (NOT directly from interface)

#### 3. PredictionModelBuilder Integration (REQUIRED)
- [ ] Add private field: `private I{FeatureName}<T>? _{featureName};` to `PredictionModelBuilder.cs`
- [ ] Add Configure method taking ONLY interface (no parameters):
  ```csharp
  public IPredictionModelBuilder<T, TInput, TOutput> Configure{FeatureName}(I{FeatureName}<T> {featureName})
  {
      _{featureName} = {featureName};
      return this;
  }
  ```
- [ ] Use feature in `Build()` with default: `var {featureName} = _{featureName} ?? new Default{FeatureName}<T>();`
- [ ] Verify feature is ACTUALLY USED in execution flow

#### 4. Beginner-Friendly Defaults (REQUIRED)
- [ ] Constructor parameters with defaults from research/industry standards
- [ ] Document WHY each default was chosen (cite papers/standards)
- [ ] Validate parameters and throw `ArgumentException` for invalid values

#### 5. Property Initialization (CRITICAL)
- [ ] NEVER use `default!` operator
- [ ] String properties: `= string.Empty;`
- [ ] Collections: `= new List<T>();` or `= new Vector<T>(0);`
- [ ] Numeric properties: appropriate default or `NumOps.Zero`

#### 6. Class Organization (REQUIRED)
- [ ] One class/enum/interface per file
- [ ] ALL interfaces in `src/Interfaces/` (root level)
- [ ] Namespace mirrors folder structure (e.g., `src/Regularization/` → `namespace AiDotNet.Regularization`)

#### 7. Documentation (REQUIRED)
- [ ] XML documentation for all public members
- [ ] `<b>For Beginners:</b>` sections with analogies and examples
- [ ] Document all `<param>`, `<returns>`, `<exception>` tags
- [ ] Explain default value choices

#### 8. Testing (REQUIRED)
- [ ] Minimum 80% code coverage
- [ ] Test with multiple numeric types (double, float)
- [ ] Test default values are applied correctly
- [ ] Test edge cases and exceptions
- [ ] Integration tests for PredictionModelBuilder usage

---

**⚠️ Failure to follow these requirements will result in repeated implementation mistakes and PR rejections.**

**See full details:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
